source/system/structs/HashMap.ooc

```diff
@@ -95,262 +60,149 @@ acX31Hash: func <K> (key: K) -> SizeT {
-				memcpy(entry value, value, V size)
-
-				entry next = null
-
-				this buckets[hash] = entry
+				prev := entry
+				entry = entry _next
+				while (entry != null && !this _keyEquals(key, entry _key)) {
+					prev = entry
+					entry = entry _next
+				}
+				if (entry != null) {
+					prev _next = entry _next
+					entry _next = null
+					result = entry _value
+					entry free()
+					this _count -= 1
+				}
 			}
-			this _count += 1
-			if ((this _count as Float / this capacity as Float) > 0.75)
-				this resize(this _count * (this _count > 50000 ? 2 : 4))
 		}
-	}
-	get: func (key: K) -> V {
-		entry: HashEntry
-		result: V = null
-		if (this getEntry(key, entry&))
-			result = entry value as V
 		result
 	}
 	contains: func (key: K) -> Bool {
-		this getEntry(key, null)
+		hashkey := This hash(key) % this capacity
+		entry := this _buckets[hashkey]
+		while (entry != null && !this _keyEquals(key, entry _key))
+			entry = entry _next
+		entry != null
+	}
+	each: func (action: Func (V*)) {
+		for (index in 0 .. this capacity) {
+			entry := this _buckets[index]
+			while (entry != null) {
+				action(entry _value&)
+				entry = entry _next
+			}
+		}
 	}
-	remove: func (key: K) -> Bool {
-		hash : SizeT = this hashKey(key) % this capacity
-		prev: HashEntry* = null
-		entry: HashEntry* = (this buckets data as HashEntry*)[hash]&
-
-		result := false
-		if (entry@ key != null) {
-			while (true) {
-				if (This _keyEquals(entry@ key as K, key)) {
-					memfree(entry@ key)
-					memfree(entry@ value)
-
-					if (prev)
-						prev@ next = entry@ next
-					else if (entry@ next)
-						this buckets[hash] = entry@ next@
-					else
-						this buckets[hash] = HashEntry new()
-
-					for (i in 0 .. this keys count) {
-						cKey := this keys[i]
-						if (This _keyEquals(key, cKey)) {
-							this keys removeAt(i)
-							break
-						}
-					}
-					this _count -= 1
-					result = true
-					break
-				}
-				if (entry@ next) {
-					prev = entry
-					entry = entry@ next
-				} else
-					break
+	each: func ~withKeys (action: Func (K*, V*)) {
+		for (index in 0 .. this capacity) {
+			entry := this _buckets[index]
+			while (entry != null) {
+				action(entry _key&, entry _value&)
+				entry = entry _next
 			}
 		}
-		result
 	}
-	resize: func (_capacity: SizeT) {
+	resize: func (newCapacity: Int) {
 		oldCapacity := this capacity
-		oldBuckets := this buckets
-		oldKeys := this keys copy()
-		this keys clear()
+		oldBuckets := this _buckets
 		this _count = 0
-		this capacity = _capacity
-		this buckets = HashEntry[this capacity] new()
+		this _capacity = newCapacity
+		this _buckets = _HashEntry[this capacity] new()
 		for (i in 0 .. oldCapacity) {
 			entry := oldBuckets[i]
-			if (entry key != null) {
-				this put(entry key as K, entry value as V)
-				old := entry
-				while (entry next) {
-					entry = entry next@
-					this put(entry key as K, entry value as V)
-				}
-				old free()
+			while (entry != null) {
```

Comment:
No, since `hash % newCapacity` will not follow any pattern similar to `hash % oldCapacity`, so the items will be organized completely differently in the "new" buckets.